Grid search color plot viz (WIP)

pbs929 opened the pull request:
Don't mean to inundate you all with PRs, but I'm having fun :)

I put together a color plot visualizer for grid search - see the ipynb for a demo.  Also aiming to create a structure into which other grid search visualizers can fit.  

Unit tests are definitely needed... I will come back to this after getting some experience with the testing workflow on my other PRs.  General feedback on the approach is welcome in the meantime. 

Called for in issue #12.

bbengfort commented:
@NealHumphrey do you have time to take a look at this by any chance? 

@pbs929 keep them coming!

bbengfort commented:
@pbs929 still interested in finishing this PR? I'm happy to do a review.

pbs929 commented:
Yes, still interested!  As I recall I was waiting on some general feedback on the approach (see demo in the ipynb).

bbengfort commented on yellowbrick/gridsearch/pcolor.py:
```diff
@@ -0,0 +1,225 @@
+    param_1 : string
+        The name of the parameter to be visualized on the horizontal axis.
+
+    param_2 : string
```
I can see the motivation for `param_1` vs `param_2` - but because these names are connected to the vertical and horizontal axes, maybe it would be easier to name them "x_param" and "y_param" - I know this is not really the x and y axes but they do it on bar graphs so I think it would be ok here.

bbengfort commented on yellowbrick/gridsearch/pcolor.py:
```diff
@@ -0,0 +1,225 @@
+    param_1 : string
+        The name of the parameter to be visualized on the horizontal axis.
+
+    param_2 : string
```
I can see the motivation for `param_1` vs `param_2` - but because these names are connected to the vertical and horizontal axes, maybe it would be easier to name them "x_param" and "y_param" - I know this is not really the x and y axes but they do it on bar graphs so I think it would be ok here.

pbs929 replied:
sure, easy enough

bbengfort commented on yellowbrick/gridsearch/pcolor.py:
```diff
@@ -0,0 +1,225 @@
+    """
+    # Get unique values of the two display parameters
+    x_vals = sorted(list(set(cv_results['param_' + param_1].compressed())))
+    y_vals = sorted(list(set(cv_results['param_' + param_2].compressed())))
```
Ok, so just to make sure I understand what's happening here - you're fetching the param mask from the cv results dictionary, compressing it to get the real values instead, then calling set to find all unique values. These are the categories for the grid, right? 

I've got a couple of notes:

1. `'param_' + param_1` is used twice, it might be easier to store these in variables and note what they are or just fetch the mask array from the results and store that. 
2. If the parameter doesn't exist in the grid search, this will return a `KeyError` - it would be better if we raised a yellowbrick key error, e.g. `YellowbrickKeyError(YellowbrickError, KeyError)` that says "parameter not found in grid search" so it's more obvious to the user what happened.

pbs929 replied:
Yes, that's right - I've added a couple comment lines to try to clarify and stored the masked array like you suggested in (1).  Also did what you suggested in (2).

bbengfort commented on yellowbrick/gridsearch/pcolor.py:
```diff
@@ -0,0 +1,225 @@
+    # This is a n_x by n_y array of lists with `None` in place of empties
+    # (my kingdom for a dataframe...)
+    all_scores = [[None for _ in range(n_x)] for _ in range(n_y)]
+    for x, y, score in zip(idx_x, idx_y, cv_results['mean_test_score']):
```
It feels like we should allow the user to choose which metric as an input to the visualizer (which defaults to `mean_test_score` but could also be - `mean_train_score`, `mean_fit_time`, and `mean_score_time`.

bbengfort commented on yellowbrick/gridsearch/pcolor.py:
```diff
@@ -0,0 +1,225 @@
+        if x is not None and y is not None:
+            if all_scores[y][x] is None:
+                all_scores[y][x] = []
+            all_scores[y][x].append(score)
```
I certainly understand the motivation for creating this intermediate data structure, but is it not possible to compute max score at this point?

bbengfort commented on yellowbrick/gridsearch/pcolor.py:
```diff
@@ -0,0 +1,225 @@
+## Dimension reduction utility
+##########################################################################
+
+def param_projection(cv_results, param_1, param_2):
```
This function seems extremely useful for all `GridSearchVisualizers` - perhaps this should be in `base.py` -- possibly as a method of `GridSearchVisualizer`?

bbengfort replied:
Or as a standalone function, with an aliased (static?) method in `GridSearchVisualizer`?

bbengfort replied:
Also: future work - add more params?

pbs929 replied:
Alright, I moved it up to the base module and aliased it in the base GridSearchVisualizer class.  I didn't make the aliased method static, rather letting it access the gridsearch model object as `self.estimator`.

Definitely could add more params as future work.

bbengfort commented on yellowbrick/gridsearch/base.py:
```diff
@@ -0,0 +1,59 @@
+## Base Grid Search Visualizer
+##########################################################################
+
+class GridSearchVisualizer(ModelVisualizer):
```
Very good set up of this into the Yellowbrick class hierarchy!

pbs929 replied:
👍

bbengfort commented on yellowbrick/gridsearch/base.py:
```diff
@@ -0,0 +1,59 @@
+        Should return an error if it isn't.
+        """
+        # A bit of type checking
+        if not is_gridsearch(model):
```
Nice!

bbengfort commented on yellowbrick/gridsearch/pcolor.py:
```diff
@@ -0,0 +1,225 @@
+        )
+
+        # Mask nans so that they can be filled with a hatch
+        data = np.ma.masked_invalid(best_scores)
```
I like this a lot! I'm going to have to use this in other things! The hatch nans thing I mean.

bbengfort commented on yellowbrick/gridsearch/pcolor.py:
```diff
@@ -0,0 +1,225 @@
+        self.ax.set_yticklabels(y_vals, rotation=45)
+
+        # Add the colorbar
+        cb = self.ax.figure.colorbar(mesh, None, self.ax)
```
This is a running debate that I hope you might weigh in on: should color bars have the range (min, max) or should they have the range (0, 1)? Should this be an option we supply to the user?

Arguments for (min, max):

- they create greater color disparity so it's easier to tell values apart 
- They bound the area you're looking in 

Arguments for (0,1):

- You can easily compare multiple heatmaps with each other 
- The closer min is to max, the more similar they appear in this as opposed to the other.

bbengfort replied:
@ndanielsen @rebeccabilbro -- just wanted to highlight this discussion for you again here.

pbs929 replied:
In this case I'd advocate for min, max.  With grid search, you can use any of a number of error metrics, which may or may not be bounded by 1.  Furthermore when fine tuning parameters you may be optimizing over the 3rd or 4th decimal of the error metric, so fixing the range on a larger scale would make the visualization not very useful.

bbengfort commented on yellowbrick/gridsearch/pcolor.py:
```diff
@@ -0,0 +1,225 @@
+        cb = self.ax.figure.colorbar(mesh, None, self.ax)
+        cb.outline.set_linewidth(0)
+
+        self.ax.set_aspect("equal")
```
Nice - glad you included this!

pbs929 replied:
👍

pbs929 commented:
Alright @bbengfort I think I've addressed your comments.  Let me know if you catch anything else - I'm pretty tied up these days with a newborn but I will try to find the time.  :)
